app.ts

```diff
@@ -9,12 +17,56 @@ const app: express.Application = express();
 
 const port: number = process.env.port ? parseInt(process.env.port) : 8080;
 
+// The OAuth code is based on multiple online sources.
+// The main one is:
+// https://www.pveller.com/oauth2-with-passport-10-steps-recipe/
+
+passport.serializeUser((user:any, done:any) => done(null, user));
+passport.deserializeUser((user:any, done:any) => done(null, user));
+
+app.use(session({
+  secret: 'keyboard cat',
+  resave: false,
+  saveUninitialized: true,
+  cookie: { secure: true }
+}));
+app.use(cors());
+app.use(bodyParser.urlencoded({ extended: false }));
 app.use(express.static('public'));
 app.use(express.json());
+app.set('trust proxy', 1) // trust first proxy
+app.use(passport.initialize());
+app.use(passport.session());
+passport.use(new OAuth2Strategy({
+    authorizationURL: 'https://signon.integration.publishing.service.gov.uk/oauth/authorize',
+    tokenURL: 'https://signon.integration.publishing.service.gov.uk/oauth/access_token',
+    clientID: process.env.OAUTH_ID,
+    clientSecret: process.env.OAUTH_SECRET,
+    callbackURL: "https://govgraphsearchdev.dev/auth/gds/callback"
+  },
+  function(accessToken: string, refreshToken: string, profile: any, cb: any) {
+    console.log('OAuth2Strategy callback', accessToken, refreshToken, profile);
+    cb(null, profile);
+  }
+));
+
+
+app.get('/login', passport.authenticate('oauth2'));
+
+
+app.get('/',
+  ensureLoggedIn('/login'),
+  async (req, res) => res.sendFile('views/index.html', {root: __dirname })
+);
+
+app.get('/auth/gds/callback',
+  passport.authenticate('oauth2', '/error-callback'),
```

Comment:
It's a mystery why this extra `passport.authenticate()` is needed here, but if it isn't here, then there's an infinite redirect loop between here and `/`.  (Max's explanation)